src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+              "Syncing archival STM from offset {} to offest in term {}",
+              last_applied_offset,
+              _last_replicated_offset);
+
+            const auto deadline = model::timeout_clock::now() + timeout;
+            const bool synced = co_await wait_no_throw(
+              *_last_replicated_offset, deadline);
+
+            if (!synced) {
+                vlog(
+                  _logger.warn,
+                  "Failed to sync archival STM from offset {} to offest {} in "
+                  "term {}",
+                  last_applied_offset,
+                  _last_replicated_offset,
+                  current_term);
+                co_return false;
+            }
+        }
+
+        if (!_raft->is_leader()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but leadership moved",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term);
+            co_return false;
+        }
+
+        if (current_term != _raft->term()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but term changed to {}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              _raft->term());
+            co_return false;
+        }
+
+        if (
+          _last_replicated_offset
+          && _manifest->get_insync_offset() < _last_replicated_offset) {
+            vlog(
+              _logger.warn,
+              "Archival STM sync from offset {} to offest in {}"
+              "term {} was successful, but the last replicated offset moved to "
+              "{}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              *_last_replicated_offset);
+            co_return false;
+        }
+
+        co_return true;
+    } else {
+        vlog(
+          _logger.debug,
+          "Syncing archival STM to latest term {}",
+          _raft->term());
+        const bool synced = co_await persisted_stm::sync(timeout);
+
+        if (!synced) {
+            vlog(_logger.debug, "Failed to sync archival STM to latest term");
+            co_return false;
+        }
+
+        if (!_raft->is_leader()) {
+            vlog(
+              _logger.debug,
+              "Archival STM sync to latest term was successful, but leadership "
+              "moved");
+            co_return false;
+        }
+
+        if (_insync_term != _raft->term()) {
+            vlog(
+              _logger.debug,
+              "Archival STM sync to latest term {} was successful, but the "
+              "term changed {}",
+              _insync_term,
+              _raft->term());
+            co_return false;
+        }
+
+        if (
+          _last_replicated_offset
```

Comment:
In this case, shouldn't we reset the replicated offset? Even though we've attempted to replicate in a previous term, there's no expectation for this STM to actually have applied it, because it's not guaranteed to have become Raft committed.

Reply:
Actually I might be wrong about this -- probably worth double checking with someone from the replication team what the guarantee is here on success of `replicate()` if you haven't already